### ISSUE TITLE
typing.ParamSpec ignored

### ISSUE THREAD

**Issue by kurtbrose** [microsoft/pyright#5738]
**Describe the bug**
Pyright does not apply `typing.ParamSpec` to return of decorator.

This seems like a correct combination of `typing` objects.   It's almost exactly parallel to the motivating examples from PEP-612 (https://peps.python.org/pep-0612/#motivation).
PyCharm supports it, mypy does not :shrug:

**Code or Screenshots**

```python
import typing

_MatchedParams = typing.ParamSpec("_MatchedParams")

_MatchFunc = typing.Callable[_MatchedParams, object]
_MatchDecorator = typing.Callable[[_MatchFunc], _MatchFunc]


def same_params_as(_matching_func: _MatchFunc) -> _MatchDecorator:
    """For the purposes of static type checking, mark a function as having the same param spec as another."""
    def decorator(func: _MatchFunc) -> _MatchFunc:
        """do-nothing decorator, just for type checking"""
        def inner(*args: _MatchedParams.args, **kwargs: _MatchedParams.kwargs):
            return func(*args, **kwargs)
        return inner
    return decorator


# test
def a(a): pass


@same_params_as(a)
def b(*args, **kwargs):
    pass


b()  # this should be highlighted as an error because it is missing a parameter; pycharm does this, vscode does not
```

here's pycharm flagging the call
![image](https://github.com/microsoft/pyright/assets/543934/10e02a6a-55b2-43d7-8918-2d9296097618)

**VS Code extension or command-line**
VSCode extension

**Comment by erictraut** [comment]
Your code has a bug in it. You are defining a generic type alias called `_MatchedParams` that is parameterized by a ParamSpec. When you use this type alias, you need to provide a corresponding type argument corresponding to the type parameter. If you omit a type argument, a type checker will assume you mean `...` (which is the ParamSpec equivalent of `Any`). That's probably not what you mean in this case.

If you enable the `reportMissingTypeArgument` diagnostic check, pyright will tell you about bugs like this. You can enable this check for your entire project by specifying it in your `pyrightconfig.json` or `pyproject.toml` file, or you can enable it locally within a file by adding the comment `# pyright: reportMissingTypeArgument=true` at the top.

Here's a corrected version of your code. Note that I've also added an explicit `TypeAlias` annotation, although you could omit this if you'd prefer.

```python
from typing import Callable, ParamSpec, TypeAlias

# pyright: reportMissingTypeArgument=true

_P = ParamSpec("_P")

_MatchFunc: TypeAlias = Callable[_P, object]
_MatchDecorator: TypeAlias = Callable[[_MatchFunc[_P]], _MatchFunc[_P]]

def same_params_as(_matching_func: _MatchFunc[_P]) -> _MatchDecorator[_P]:
    """For the purposes of static type checking, mark a function as having the same param spec as another."""

    def decorator(func: _MatchFunc[_P]) -> _MatchFunc[_P]:
        """do-nothing decorator, just for type checking"""

        def inner(*args: _P.args, **kwargs: _P.kwargs):
            return func(*args, **kwargs)

        return inner

    return decorator
```

I'm not sure what PyCharm is doing here, but it's definitely not following the official typing specification.

**Comment by kurtbrose** [comment]
Thanks!!  Sorry for the waste of time but that was a very helpful response!

**Comment by kurtbrose** [comment]
![image](https://github.com/microsoft/pyright/assets/543934/fc0dc7b6-2431-41ea-b77e-c6322770970d)
:-) thanks again!

**Comment by erictraut** [comment]
Happy to help. This is a very common point of confusion when using type aliases. That's one of the main motivations for [PEP 695](https://peps.python.org/pep-0695/), which includes new dedicated syntax for defining type aliases. It's included in the upcoming Python 3.12 release. The new syntax should make things much clearer.
